geonode/layers/utils.py

```diff
@@ -457,7 +457,7 @@ def file_upload(filename,
         if latest_uploads.count() > 1:
             upload_session = latest_uploads.first()
         else:
-            upload_session, _created = UploadSession.objects.get_or_create(resource=layer)
+            upload_session, _created = UploadSession.objects.get_or_create(resource=layer, defaults={'user': theuser})
         upload_session.user = theuser
```

Comment:
we should probably remove this line too then.

Reply:
I thought about this, but it depends on if the code should alter the user of an existing Upload Session?

Reply:
If I'm correctly interpreting the code, it looks like the intention was to update the Upload Session user in any case, but probably we are missing an `upload_session.save()` after the assignment `upload_session.user = theuser`.

In practice the associated user to an Upload Session will be used only to show the incomplete/pending uploads on the Upload page. A `superuser` will be able to see all of them anyway. So IMHO we should always update the user and make sure it will be correctly saved later on.

Reply:
Should the upload session be immediately saved if it's just been created?

My assumption is that the `latest_uploads` object is retrieved because there may have been multiple upload sessions for the same resource (layer in this case).

I think there's 3 cases here:
 1. Multiple uploads exist already, and we want to reuse the first one (most recent)
 2. A single upload session exists and we want to use that one
 3. No upload session exists, so we want to create a new one
 
 The code might be better off looking something like this:
 
 ```
 latest_uploads = UploadSession.objects.filter(resource=layer).order_by('-date')
 if latest_uploads.count() >= 1:
     upload_session = latest_uploads.first()
     upload_session.user = theuser
 else:
     upload_session = UploadSession.objects.create(resource=layer, user=theuser)
```

It also makes it easier to do an `upload_session.save()` after the user update.

Let me know what you think @afabiani :)

Reply:
@alistairjmcintyre yes, I agree with your solution. Please proceed! Thanks :)

Reply:
Awesome. I've pushed the change. I chose to omit the `save()` call as one is made further down the same method. 

I also swapped the `.count()` to a `len()` call as this won't do a `SELECT COUNT(*) ...` database query, instead just checking how many objects in the python iterable instead.